### PR TITLE
Introduce `ProviderFactory#zip`

### PR DESCRIPTION
**Pull request by melix**
This PR introduces a `zip` method which allows combining two providers into a new live provider. The dependencies of both providers are tracked (because the resulting provider _depends on_ both values).

See #12974

### CODE REVIEW

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test this

**Comment by melix** [comment]
@bot-gradle test RFM

**Comment by melix** [comment]
@adammurdoch any thoughts on this one?

File: subprojects/core-api/src/main/java/org/gradle/api/provider/ProviderFactory.java [hunk 1]
```diff
@@ -232,4 +233,24 @@
...
+     * @param combiner the combiner of values
+     * @param <A> the type of the first provider
+     * @param <B> the type of the second provider
+     * @param <R> the type of the result of the combiner
+     * @return a combined provider
+     *
+     * @since 6.6
+     */
+    @Incubating
+    <A, B, R> Provider<R> zip(Provider<A> left, Provider<B> right, BiFunction<A, B, R> combiner);
```
Comment (big-guy) [inline subprojects/core-api/src/main/java/org/gradle/api/provider/ProviderFactory.java:255]:
```suggestion
    <A, B, R> Provider<R> zip(Provider<A> first, Provider<B> second, BiFunction<A, B, R> combiner);
```

Comment (big-guy) [reply subprojects/core-api/src/main/java/org/gradle/api/provider/ProviderFactory.java:255]: I don't feel too strongly about this

**Pull request merged**